// swagger-client/lib/resolver/apidom/reference/dereference/strategies/openapi-3-1-swagger-client/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
var _empty = require("@swagger-api/apidom-reference/configuration/empty");
var _openapi = _interopRequireDefault(require("@swagger-api/apidom-reference/dereference/strategies/openapi-3-1"));
var _apidomNsOpenapi = _interopRequireWildcard(require("@swagger-api/apidom-ns-openapi-3-1"));
var _dereference = _interopRequireDefault(require("./visitors/dereference.js"));
var _parameters = _interopRequireDefault(require("./visitors/parameters.js"));
var _properties = _interopRequireDefault(require("./visitors/properties.js"));
var _allOf = _interopRequireDefault(require("./visitors/all-of.js"));
/* eslint-disable camelcase */

const visitAsync = _apidomCore.visit[Symbol.for('nodejs.util.promisify.custom')];
const mergeAllVisitorsAsync = _apidomCore.mergeAllVisitors[Symbol.for('nodejs.util.promisify.custom')];
const OpenApi3_1SwaggerClientDereferenceStrategy = _openapi.default.compose({
  props: {
    allowMetaPatches: false,
    parameterMacro: null,
    modelPropertyMacro: null,
    mode: 'non-strict',
    ancestors: null
  },
  init({
    allowMetaPatches = this.allowMetaPatches,
    parameterMacro = this.parameterMacro,
    modelPropertyMacro = this.modelPropertyMacro,
    mode = this.mode,
    ancestors = []
  } = {}) {
    this.name = 'openapi-3-1-swagger-client';
    this.allowMetaPatches = allowMetaPatches;
    this.parameterMacro = parameterMacro;
    this.modelPropertyMacro = modelPropertyMacro;
    this.mode = mode;
    this.ancestors = [...ancestors];
  },
  methods: {
    async dereference(file, options) {
      var _options$dereference$;
      const visitors = [];
      const namespace = (0, _apidomCore.createNamespace)(_apidomNsOpenapi.default);
      const immutableRefSet = (_options$dereference$ = options.dereference.refSet) != null ? _options$dereference$ : (0, _empty.ReferenceSet)();
      const mutableRefsSet = (0, _empty.ReferenceSet)();
      let refSet = immutableRefSet;
      let reference;
      if (!immutableRefSet.has(file.uri)) {
        reference = (0, _empty.Reference)({
          uri: file.uri,
          value: file.parseResult
        });
        immutableRefSet.add(reference);
      } else {
        // pre-computed refSet was provided as configuration option
        reference = immutableRefSet.find(ref => ref.uri === file.uri);
      }

      /**
       * Clone refSet due the dereferencing process being mutable.
       * We don't want to mutate the original refSet and the references.
       */
      if (options.dereference.immutable) {
        immutableRefSet.refs.map(ref => (0, _empty.Reference)({
          ...ref,
          value: (0, _apidomCore.cloneDeep)(ref.value)
        })).forEach(ref => mutableRefsSet.add(ref));
        reference = mutableRefsSet.find(ref => ref.uri === file.uri);
        refSet = mutableRefsSet;
      }

      // create main dereference visitor
      const dereferenceVisitor = (0, _dereference.default)({
        reference,
        namespace,
        options,
        allowMetaPatches: this.allowMetaPatches,
        ancestors: this.ancestors
      });
      visitors.push(dereferenceVisitor);

      // create parameter macro visitor (if necessary)
      if (typeof this.parameterMacro === 'function') {
        const parameterMacroVisitor = (0, _parameters.default)({
          parameterMacro: this.parameterMacro,
          options
        });
        visitors.push(parameterMacroVisitor);
      }

      // create model property macro visitor (if necessary)
      if (typeof this.modelPropertyMacro === 'function') {
        const modelPropertyMacroVisitor = (0, _properties.default)({
          modelPropertyMacro: this.modelPropertyMacro,
          options
        });
        visitors.push(modelPropertyMacroVisitor);
      }

      // create allOf visitor (if necessary)
      if (this.mode !== 'strict') {
        const allOfVisitor = (0, _allOf.default)({
          options
        });
        visitors.push(allOfVisitor);
      }

      // establish root visitor by visitor merging
      const rootVisitor = mergeAllVisitorsAsync(visitors, {
        nodeTypeGetter: _apidomNsOpenapi.getNodeType
      });
      const dereferencedElement = await visitAsync(refSet.rootRef.value, rootVisitor, {
        keyMap: _apidomNsOpenapi.keyMap,
        nodeTypeGetter: _apidomNsOpenapi.getNodeType
      });

      /**
       * If immutable option is set, replay refs from the refSet.
       */
      if (options.dereference.immutable) {
        mutableRefsSet.refs.filter(ref => ref.uri.startsWith('immutable://')).map(ref => (0, _empty.Reference)({
          ...ref,
          uri: ref.uri.replace(/^immutable:\/\//, '')
        })).forEach(ref => immutableRefSet.add(ref));
        reference = immutableRefSet.find(ref => ref.uri === file.uri);
        refSet = immutableRefSet;
      }

      /**
       * Release all memory if this refSet was not provided as an configuration option.
       * If provided as configuration option, then provider is responsible for cleanup.
       */
      if (options.dereference.refSet === null) {
        immutableRefSet.clean();
      }
      mutableRefsSet.clean();
      return dereferencedElement;
    }
  }
});
var _default = exports.default = OpenApi3_1SwaggerClientDereferenceStrategy;
/* eslint-enable camelcase */
